eoscompanion/content_adjusters.py

```diff
@@ -18,46 +18,283 @@
 
-    Right now the way that this is done is a total hack (regex), in future
-    we might want to depend on beautifulsoup and use that instead.
+    If :ekn_uri: is None, return None.
     '''
-    return EosCompanionAppService.string_to_bytes(
-        _RE_EKN_URL_CAPTURE.sub(ekn_url_rewriter(query),
-                                EosCompanionAppService.bytes_to_string(content_bytes))
-    )
+    if ekn_uri is None:
+        return None
+
+    return rewrite_ekn_url(os.path.basename(urlparse(ekn_uri).path), query)
+
+
+_MOBILE_WRAPPER_TEMPLATE_URI = (
+    'resource:///com/endlessm/CompanionAppService/data/templates/mobile-article-wrapper.mst'
+)
+
+
+def render_mobile_wrapper(renderer,
+                          app_id,
+                          rendered_content,
+                          metadata,
+                          engine,
+                          shards,
+                          query,
+                          callback):
+    '''Render the page wrapper and initialize crosslinks.
+
+    This is the final rendering step that EKN content needs to undergo before
+    it becomes usable, at least on the desktop. This step initializes
+    both a dictionary of metadata about the content and also a map of
+    outgoing links to internal links. Then we inject some javascript which
+    at browser-render time, rewrites the page to resolve all those links.
+    '''
+    def _on_queried_sets(src, result):
+        '''Called when we finish querying set objects.'''
+        try:
+            results = src.query_finish(result)
+        except GLib.Error as error:
+            callback(error, None)
+            return
+
+        set_objects = results.get_property('models')
+        sets_for_tags = {
+            tag: set_object
+            for tag, set_object in tagged_sets(set_objects)
+            if not tag.startswith('Ekn')
+        }
+        content_metadata = {
+            'title': metadata.get('title', ''),
+            'published': metadata.get('published', ''),
+            'authors': metadata.get('authors', []),
+            'license': metadata.get('license', ''),
+            'source': metadata.get('source', ''),
+            'source_name': metadata.get('sourceName', ''),
+            'originalURI': metadata.get('originalURI', ''),
+            'sets': [
+                {
+                    'child_tags': set_object.get_property('child-tags'),
+                    'ekn_id': set_object.get_property('ekn-id'),
+                    'title': set_object.get_property('title'),
+                    'tags': set_object.get_property('tags')
+                }
+                for set_object in
+                yield_set_objects_for_tags(sets_for_tags,
+                                           metadata.get('tags', []))
+            ]
+        }
+
+        # Now that we have everything, read the template and render it
+        template_file = Gio.File.new_for_uri(_MOBILE_WRAPPER_TEMPLATE_URI)
```

Comment:
You could move this inside the `try` block below

Reply:
`g_file_new_for_uri` does not take a `GError **` https://developer.gnome.org/gio/stable/GFile.html#g-file-new-for-uri

Reply:
My point here was that since we don't use `template_file` anywhere else, it would make the code a bit more neat to have it there. But up to you.

Reply:
My preference would be to minimize the amount of code in `try` blocks so that we know exception specifiers are catching the right thing.

Reply:
I'm fine leaving it as-is, but in this case does it make a difference though, since we know the `GFile` function does not throw?